src/mission_companion.cpp

```diff
@@ -1284,203 +3053,1208 @@ bool talk_function::forage_return( npc &p )
+            }
+        }
+        std::vector<std::shared_ptr<npc>> remaining_def;
+        for( const auto &elem : defender ) {
+            if (elem->hp_cur[hp_torso] != 0){
+                remaining_def.push_back(elem);
+            }
+        }
+
+        defense = combat_score(remaining_def);
+        attack = combat_score(remaining_att);
+        if (attack > defense*3){
+            attack_random( remaining_att, remaining_def);
+            if (defense == 0 || (remaining_def.size() == 1 && remaining_def[0]->hp_cur[hp_torso] == 0 )){
+                popup(_("%s forces are destroyed!"), defender[0]->my_fac->name.c_str());
+            } else {
+                popup(_("%s forces retreat from combat!"), defender[0]->my_fac->name.c_str());
+            }
+            return;
+        } else if (attack*3 < defense) {
+            attack_random( remaining_def, remaining_att);
+            if (attack == 0 || (remaining_att.size() == 1 && remaining_att[0]->hp_cur[hp_torso] == 0 )){
+                popup(_("%s forces are destroyed!"), attacker[0]->my_fac->name.c_str());
+            } else {
+                popup(_("%s forces retreat from combat!"), attacker[0]->my_fac->name.c_str());
+            }
+            return;
+        } else {
+            def_init = rng( 1, 6 ) + advantage;
+            att_init = rng( 1, 6 );
+            if (def_init >= att_init) {
+                attack_random( remaining_att, remaining_def);
+            }
+            if (def_init <= att_init) {
+                attack_random( remaining_def, remaining_att);
+            }
+        }
+    }
+}
+
+void talk_function::companion_return( npc &comp ){
+    assert( !comp.is_active() );
+    comp.reset_companion_mission();
+    comp.companion_mission_time = calendar::before_time_starts;
+    comp.companion_mission_time_ret = calendar::before_time_starts;
+    for( size_t i = 0; i < comp.companion_mission_inv.size(); i++ ) {
+        for( const auto &it : comp.companion_mission_inv.const_stack( i ) ) {
+            g->m.add_item_or_charges( g->u.posx(), g->u.posy(), it );
+        }
+    }
+    comp.companion_mission_inv.clear();
+    comp.companion_mission_points.clear();
+    // npc *may* be active, or not if outside the reality bubble
+    g->reload_npcs();
+}
+
+std::vector<std::shared_ptr<npc>> talk_function::companion_list( const npc &p, const std::string &id, bool contains )
+{
+    std::vector<std::shared_ptr<npc>> available;
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( p.posx(), p.posy() ) );
+    for( const auto &elem : overmap_buffer.get_companion_mission_npcs() ) {
+        npc_companion_mission c_mission = elem->get_companion_mission();
+        if( c_mission.position == tripoint( omt_pos.x, omt_pos.y, p.posz() ) &&
+           c_mission.mission_id == id && c_mission.role_id == p.companion_mission_role_id ){
+
+            available.push_back( elem );
+        } else if( contains && c_mission.mission_id.find( id ) != std::string::npos ){
```

Comment:
Don't store this as a substring. Add this to the struct, as a totally separate struct.
Same for direction - it certainly shouldn't be encoded as a string. A best candidate for direction type would be `om_direction::type` from `omdata.h` - it already has all the functions like getting name, offsetting a point by it, opposite direction etc.